src/Hamiltonians/ho-cart-tools.jl

```diff
@@ -0,0 +1,203 @@
+* `block_id`: index of block in order found
+* `block_E0`: noninteracting energy of all elements in the block
+* `block_size`: number of elements in the block
+* `addr`: first address that generates the block with e.g. [`build_basis`](@ref)
+* `indices`: tuple of mode indices that allow recreation of the generating address 
+    `addr`; in this case use e.g. `BoseFS(M; indices .=> 1)` This is useful when 
+    the `DataFrame` is loaded from file since [`Arrow.jl`](@ref) converts custom
+    types to `NamedTuple`s.
+* `t_basis`: time to generate the basis for each block
+
+Keyword arguments:
+* `target_energy`: only blocks with this noninteracting energy are found
+* `max_energy`: only blocks with noninteracting energy less than this are found
+* `max_blocks`: exit after finding this many blocks
+* `method`: Choose between `:vertices` and `:comb` for method of enumerating tuples of quantum numbers
+* `save_to_file=nothing`: if set then the `DataFrame` recording blocks is saved after each new block is found
+* additional `kwargs`: passed to `isapprox` for comparing block energies. Useful for anisotropic traps
+"""
+function get_all_blocks(h::Union{HOCartesian{D},HOCartesianSeparable{D}}; 
+        target_energy = nothing, 
+        max_energy = nothing, 
+        max_blocks = nothing, 
+        method = :vertices,
+        kwargs...
+    ) where {D}
+
+    add0 = starting_address(h)
+    N = num_particles(add0)
+    E0 = N * sum(h.aspect1) / 2  # starting address may not be ground state
+    if !isnothing(target_energy) && target_energy - E0 > minimum(h.S .* h.aspect1) - 1
+        @warn "target energy higher than single particle basis size, not all blocks may be found."
+    end
+    if !isnothing(max_energy) && max_energy < E0
+        @warn "maximum requested energy lower than groundstate, not all blocks may be found."
+    end
+    if !isnothing(max_energy) && !isnothing(target_energy) && max_energy < target_energy
+        @warn "maximum requested energy lower than target energy, not all blocks may be found."
+    end
+
+    if h isa HOCartesian
+        M = h.S[1] - 1
+        if (isnothing(max_energy) && isnothing(target_energy)) ||
+            (!isnothing(max_energy) && max_energy > E0 + M) ||
+            (!isnothing(target_energy) && target_energy > E0 + M)
+            throw(ArgumentError("requested energy range not accessible by given Hamiltonian"))
+        end
+    end
+
+    if method == :vertices
+        df = get_all_blocks_vertices(h; target_energy, max_energy, max_blocks, kwargs...)
+    elseif method == :comb
+        df = get_all_blocks_comb(h; target_energy, max_energy, max_blocks, kwargs...)
+    else
+        throw(ArgumentError("invalid method."))
+    end
+
+    # consistency check
+    # if isnothing(max_blocks) && isnothing(target_energy) && isnothing(max_energy) && sum(df[!,:block_size]) ≠ dimension(h)
+    #     @warn "not all blocks were found"
+    # end
+    # return df
```

Comment:
```suggestion
    if method == :vertices
        return get_all_blocks_vertices(h; target_energy, max_energy, max_blocks, kwargs...)
    elseif method == :comb
        return get_all_blocks_comb(h; target_energy, max_energy, max_blocks, kwargs...)
    else
        throw(ArgumentError("invalid method."))
    end
```
Keeping the consistency check is a good option too, if there is a danger of not finding all blocks and it doesn't take too long.

Reply:
this consistency check is not needed anymore, I forgot to remove it entirely.